### PR TITLE
Projects and GitHub integration

### PR DESCRIPTION
**Pull request by linusha**

### CODE REVIEW

**Review by merryman — COMMENTED** [review]
LGTM overall. Minor adjustments needed, also will further test the whole thing in detail before giving final approval.

File: lively.components/prompts.cp.js [hunk 1]
```diff
@@ -93,6 +94,11 @@ export class InformPromptModel extends AbstractPromptModel {
     promptTitle.lineWrapping = this.lineWrapping;
     promptTitle.fixedWidth = !!this.lineWrapping;
     promptTitle.value = this.label;
+    if (this.additionalText) {
+      const additionalText = new Text();
+      additionalText.value = this.additionalText;
+      this.view.addMorph(additionalText, this.ui.okButton);
```
[the remark this answers is not in the document]
Comment (linusha) [reply lively.components/prompts.cp.js:100]: Done

File: lively.ide/world-commands.js [hunk 2]
```diff
@@ -1195,23 +1198,27 @@ const commands = [
...
+      let saved;
+      if ($world.openedProject) {
+        saved = await $world.openPrompt(part(SaveProjectDialog, { viewModel: { project: $world.openedProject } }));
+      } else { // in case there is another morph implementing save...
+        const relayed = evt && world.relayCommandExecutionToFocusedMorph(evt);
+        if (relayed) return relayed;
+        args = { confirmOverwrite: true, showSaveDialog: true, moduleManager: modules, ...args };
+        const focused = world.focusedMorph;
+        saved = await interactivelySaveWorld(world, args);
+        if (focused && focused.focus());
```
[the remark this answers is not in the document]
Comment (linusha) [reply lively.ide/world-commands.js:1212]: Removedf

File: lively.ide/studio/world-browser.cp.js [hunk 3]
```diff
@@ -637,41 +626,40 @@ class GrowingWorldList extends Morph {
...
-            closeVersionsButton: this.getSubmorphNamed('close versions button'),
-            versionList: this.getSubmorphNamed('version list'),
-            deleteWorldPrompt: this.getSubmorphNamed('delete world prompt'),
-            cancelDeleteButton: this.getSubmorphNamed('cancel delete button'),
-            confirmDeleteButton: this.getSubmorphNamed('confirm delete button')
-          };
+          return [
+            ...this.bindingsToInherit,
+            // TODO: this cannot work, this is not actually on this component but gets added later.
+            // previosly this only worked since we had a dynamic lookup based on morph name
```
[the remark this answers is not in the document]
Comment (linusha) [reply lively.ide/studio/world-browser.cp.js:645]: Checked, works now.

File: lively.ide/studio/world-browser.cp.js [hunk 4]
```diff
@@ -946,7 +905,7 @@ const WorldPreviewTile = component({
       submorphs: [{
         name: 'label',
         fontWeight: 600,
-        textAndAttributes: ['OPEN PROJECT', null]
+        textAndAttributes: ['OPEN WORLD', null]
```
[the remark this answers is not in the document]
Comment (linusha) [reply lively.ide/studio/world-browser.cp.js:908]: Has been changed already.

File: lively.morphic/config.js [hunk 5]
```diff
@@ -72,6 +72,8 @@ const config = {
 
   ide: {
 
+    projectsEnabled: false,
```
[the remark this answers is not in the document]
Comment (linusha) [reply lively.morphic/config.js:75]: Removed

File: lively.ide/world.js [hunk 6]
```diff
@@ -389,19 +406,123 @@ export class LivelyWorld extends World {
...
-      .animate({ blur: 0, duration: 1000, easing: easings.inOutExpo })
-      .then(() => {
-        document.body.style.background = 'black';
+      this
+        .animate({ blur: 0, duration: 1000, easing: easings.inOutExpo })
+        .then(() => {
+          document.body.style.background = 'black';
+        });
+      document.body.style.overflowX = 'visible';
+      document.body.style.overflowY = 'visible';
```
[the remark this answers is not in the document]
Comment (linusha) [reply lively.ide/world.js:415]: Done

File: lively.server/plugins/world-loading.js [hunk 7]
```diff
@@ -67,11 +67,12 @@ export default class WorldLoadingPlugin {
     query = query ? '?' + query : '';
 
     if ((url === '/' || url === '/index.html') && req.method.toUpperCase() === 'GET') {
-      res.writeHead(301, { location: '/worlds/' + query });
+      res.writeHead(301, { location: '/worlds/' + query }); // might get redirected further below
       res.end();
       return;
     }
 
+    // TODO: is this still used?
```
[the remark this answers is not in the document]
Comment (linusha) [reply lively.server/plugins/world-loading.js:75]: Please check that I removed only the appropriate block. Thanks.

File: lively.project/project.js [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+    const shellBuildScript = join(this.url, 'tools/build.sh');
+    if (!await resource(shellBuildScript).exists()) throw Error('build.sh not found');
+    await resource(shellBuildScript).write(buildScriptShell);
+
+    const mjsBuildScript = join(this.url, 'tools/build.mjs');
+    if (!await resource(mjsBuildScript).exists()) throw Error('build.mjs not found');
+    let content = buildScript;
+    content = content.replaceAll('%PROJECT_NAME%', this.name);
+    await resource(mjsBuildScript).write(content);
+    const scriptDir = shellBuildScript.replace('http://localhost:9011/', '').replace('/build.sh', '');
```
[the remark this answers is not in the document]
Comment (linusha) [reply lively.project/project.js:326]: Should be more resilient now:

```js
const shellBuildScriptHost = new URL(shellBuildScript).host;
const scriptDir = shellBuildScript.replace(shellBuildScriptHost, '').replace('/build.sh', '');
```

File: lively.ide/studio/world-browser.cp.js [hunk 9]
```diff
@@ -680,31 +668,29 @@ export class WorldPreview extends Morph {
...
-    this._commit = commit;
     preview.imageUrl = commit.preview || missingSVG;
     let { name: authorName } = commit.author;
-    authorName = authorName.startsWith('guest-') ? 'guest' : authorName;
+    authorName = authorName.startsWith('guest') ? 'guest' : authorName;
     timestamp.value = [authorName, { fontSize: 13, fontWeight: 'bold', paddingTop: '1px' }, date.format(commit.timestamp, ' - d.m.yy HH:MM'), {
       fontWeight: 'bold', fontSize: 12, paddingTop: '2px'
     }];
-    title.title = commit.name;
+    title.value = commit.name;
```
[the remark this answers is not in the document]
Comment (linusha) [reply lively.ide/studio/world-browser.cp.js:680]: Done!

```js
timestamp.value = [project.projectRepoOwner + ' - ' + authorName, { fontSize: 13, fontWeight: 'bold', paddingTop: '1px' }];
timestamp.tooltip = `Project ${project.name} was created by ${authorName}.\nIts repository belongs to ${project.projectRepoOwner}.`;
```

File: lively.user/user-flap.cp.js [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+    const deviceCodeMatch = resOne.match(new RegExp('device_code=(.*)&e'));
+    const userCodeMatch = resOne.match(new RegExp('user_code=(.*)&'));
+    if (!deviceCodeMatch || !userCodeMatch) {
+      $world.setStatusMessage('An unexpected error occured. Please contact the lively.next team.', StatusMessageError);
+      return;
+    }
+    const deviceCode = deviceCodeMatch[1];
+    const userCode = userCodeMatch[1];
+    const confirm = await $world.confirm(['Go to ', null, 'GitHub', { doit: { code: 'window.open(\'https://github.com/login/device\',\'Github Authentification\',\'width=500,height=600,top=100,left=500\')' }, fontColor: Color.link }, ` and enter\n${userCode}\n Afterwards, confirm with OK.`, null]);
+    if (!confirm) return;
```
[the remark this answers is not in the document]
Comment (linusha) [reply lively.user/user-flap.cp.js:57]: I don't think its that easy, but that might just be me blocking the hose :firefighter:. Let's talk about it tomorrow :)

**Review by merryman — APPROVED** [review]

**Pull request merged**